toolchain/lower/file_context.h

```diff
@@ -157,6 +188,50 @@ class FileContext {
 
   auto BuildVtable(const SemIR::Class& class_info) -> llvm::GlobalVariable*;
 
+  // Function is used to track which specifics were lowered for each generic.
+  // These are added one by one while lowering their definitions.
+  auto AddLoweredSpecificForGeneric(SemIR::GenericId generic_id,
+                                    SemIR::SpecificId specific_id) {
+    lowered_specifics_[generic_id.index].push_back(specific_id);
+  }
+  // Initialize a return a SpecificFunctionFingerprint* instance for specific.
+  // The internal of the fingerprint are populated during and after lowering
+  // the function body of that specific.
+  auto InitializeFingerprintForSpecific(SemIR::SpecificId specific_id)
+      -> SpecificFunctionFingerprint* {
+    if (!specific_id.has_value()) {
+      return nullptr;
+    }
+    return &lowered_specific_fingerprint_[specific_id.index];
+  }
+
+  // Entry point for coalescing equivalent specifics.
+  auto CoalesceEquivalentSpecifics() -> void;
+  // While coalescing specifics, compare the function types for two specifics.
+  // This uses a fingerprint generated for each function type.
+  auto AreFunctionTypesEquivalent(SemIR::SpecificId specific1,
+                                  SemIR::SpecificId specific2) -> bool;
+  // While coalescing specifics, compare the function bodies for two specifics.
+  // This uses fingerprints generated during lowering of the function body.
+  auto AreFunctionBodiesEquivalent(
+      SemIR::SpecificId specific1, SemIR::SpecificId specific2,
+      Set<std::pair<SemIR::SpecificId, SemIR::SpecificId>>&
+          visited_equivalent_specifics,
+      Set<std::pair<SemIR::SpecificId, SemIR::SpecificId>>&
+          visited_equivalent_specifics_flipped) -> bool;
+  // When an equivalence is found, update the canonical specific to use for
+  // each of the two Specifics found to be equivalent, replace all uses of one
+  // specific with the canonical one; mark one specific for deletion using
+  // specifics_to_delete.
+  auto ProcessSpecificEquivalence(
+      std::pair<SemIR::SpecificId, SemIR::SpecificId>& pair,
```

Comment:
Does this need to be a reference?

```suggestion
      std::pair<SemIR::SpecificId, SemIR::SpecificId> pair,
```